### PR TITLE
Custom class domain/specialization getters and setters

### LINKED ISSUE #742: NwNx_Creatures - Update functions for Schools/Domains (8193.6 and +)
Hi,

Since patch 8193.6, any class can now have Domains and a School of magic, but NwNx_Creatures at the moment only looks for these variables under the Cleric/Wizard classes. Could the following functions be updated to allow us to input the Class?

int NWNX_Creature_GetClericDomain(object creature, int index);
void NWNX_Creature_SetClericDomain(object creature, int index, int domain);
int NWNX_Creature_GetWizardSpecialization(object creature);
void NWNX_Creature_SetWizardSpecialization(object creature, int school);

Thanks!

### PR DESCRIPTION
**Pull request by GoLoT**
I reinstalled my OS recently and I haven't set up my testing environment yet so I haven't been able to test the changes. It's an easy fix, I don't see what could be wrong or missing but still I'm tagging it as untested. I hope I will be able to test it tomorrow once I get my server running.

Leaving it here in case someone wants to test/fix it before I get to it.

EDIT: Hopefully fixes #742

### CODE REVIEW

File: Plugins/Creature/Creature.cpp [hunk 1]
```diff
@@ -1191,6 +1195,120 @@ ArgumentStack Creature::SetWizardSpecialization(ArgumentStack&& args)
...
+        ASSERT_OR_THROW(school <= 255);
+        ASSERT_OR_THROW(school >= 0);
+
+        CNWClass* pClass = classId < Globals::Rules()->m_nNumClasses ? &Globals::Rules()->m_lstClasses[classId] : nullptr;
+        //ASSERT_OR_THROW(pClass->m_bHasSpecialization);
+
+        for (int32_t i = 0; i < 3; i++)
+        {
+            auto& classInfo = pCreature->m_pStats->m_ClassInfo[i];
+            if (classInfo.m_nClass == classId && pClass->m_bHasSpecialization)
```
Comment (mtijanic) [inline Plugins/Creature/Creature.cpp]: Don't check for `bHas{Specialization,Domains}`. If someone wants to give one specific fighter a school specialization, let's not prevent it.

File: Plugins/Creature/NWScript/nwnx_creature.nss [hunk 2]
```diff
@@ -375,12 +377,14 @@ void NWNX_Creature_SetAlignmentLawChaos(object creature, int value);
...
 /// @param creature The creature object.
 /// @param index The first or second domain.
+/// @deprecated Use NWNX_Creature_GetDomain(). This will be removed in future NWNX releases.
 int NWNX_Creature_GetClericDomain(object creature, int index);
 
 /// @brief Sets one of creature's cleric domains.
 /// @param creature The creature object.
 /// @param index The first or second domain.
 /// @param domain The domain constant to set.
+/// @deprecated Use NWNX_Creature_SetDomain(). This will be removed in future NWNX releases.
```
Comment (mtijanic) [inline Plugins/Creature/NWScript/nwnx_creature.nss]: Would be nice if this was removed from nwnx and translation done in nwscript..

File: Plugins/Creature/Creature.cpp [hunk 3]
```diff
@@ -1191,6 +1195,120 @@ ArgumentStack Creature::SetWizardSpecialization(ArgumentStack&& args)
...
+    return stack;
+}
+
+ArgumentStack Creature::GetSpecialization(ArgumentStack&& args)
+{
+    ArgumentStack stack;
+    int32_t retVal = -1;
+
+    const auto classId = Services::Events::ExtractArgument<int32_t>(args);
+    ASSERT_OR_THROW((classId >= 0) && (classId < 255));
```
Comment (mtijanic) [inline Plugins/Creature/Creature.cpp]: Use the constants like in the other functions.

File: Plugins/Creature/Creature.cpp [hunk 4]
```diff
@@ -1191,6 +1195,120 @@ ArgumentStack Creature::SetWizardSpecialization(ArgumentStack&& args)
...
+    {
+        const auto classId = Services::Events::ExtractArgument<int32_t>(args);
+        ASSERT_OR_THROW(classId >= Constants::ClassType::MIN);
+        ASSERT_OR_THROW(classId <= Constants::ClassType::MAX);
+        const auto school = Services::Events::ExtractArgument<int32_t>(args);
+        ASSERT_OR_THROW(school <= 255);
+        ASSERT_OR_THROW(school >= 0);
+
+        CNWClass* pClass = classId < Globals::Rules()->m_nNumClasses ? &Globals::Rules()->m_lstClasses[classId] : nullptr;
+        //ASSERT_OR_THROW(pClass->m_bHasSpecialization);
```
Comment (mtijanic) [inline Plugins/Creature/Creature.cpp]: `ASSERT_OR_THROW(pClass != nullptr);` would be nice here.

**Comment by mtijanic** [comment]
Nice. I didn't test it yet though.

**Comment by GoLoT** [comment]
Changes added, thanks for the review. I will try to get my testing environment up ASAP so I can test it.

File: Plugins/Creature/NWScript/nwnx_creature.nss [hunk 5]
```diff
@@ -1092,44 +1120,22 @@ void NWNX_Creature_SetAlignmentLawChaos(object creature, int value)
...
 void NWNX_Creature_SetClericDomain(object creature, int index, int domain)
 {
-    string sFunc = "SetClericDomain";
-
-    NWNX_PushArgumentInt(NWNX_Creature, sFunc, domain);
-    NWNX_PushArgumentInt(NWNX_Creature, sFunc, index);
-    NWNX_PushArgumentObject(NWNX_Creature, sFunc, creature);
-
-    NWNX_CallFunction(NWNX_Creature, sFunc);
+    return NWNX_Creature_SetDomain(creature, CLASS_TYPE_CLERIC, index, domain);
```
Comment (Daztek) [inline Plugins/Creature/NWScript/nwnx_creature.nss]: Don't need a `return` here and could you add a `WriteTimestampedLogEntry("WARNING: NWNX_Creature_SetClericDomain() is deprecated, please use NWNX_Creature_SetDomain()");`?

File: Plugins/Creature/NWScript/nwnx_creature.nss [hunk 6]
```diff
@@ -1092,44 +1120,22 @@ void NWNX_Creature_SetAlignmentLawChaos(object creature, int value)
...
 
 void NWNX_Creature_SetWizardSpecialization(object creature, int school)
 {
-    string sFunc = "SetWizardSpecialization";
-
-    NWNX_PushArgumentInt(NWNX_Creature, sFunc, school);
-    NWNX_PushArgumentObject(NWNX_Creature, sFunc, creature);
-
-    NWNX_CallFunction(NWNX_Creature, sFunc);
+    return NWNX_Creature_SetSpecialization(creature, CLASS_TYPE_WIZARD, school);
```
Comment (Daztek) [inline Plugins/Creature/NWScript/nwnx_creature.nss]: Also doesn't need a `return` and could use: `WriteTimestampedLogEntry("WARNING: NWNX_Creature_SetWizardSpecialization() is deprecated, please use NWNX_Creature_SetSpecialization()");`

File: Plugins/Creature/Creature.cpp [hunk 7]
```diff
@@ -1149,16 +1160,23 @@ ArgumentStack Creature::SetClericDomain(ArgumentStack&& args)
...
     int32_t retVal = -1;
-    if (auto *pCreature = creature(args))
+
+    const auto classId = Services::Events::ExtractArgument<int32_t>(args);
+    ASSERT_OR_THROW((classId >= Constants::ClassType::MIN) && (classId <= Constants::ClassType::MAX));
+
+    CNWClass* pClass = classId < Globals::Rules()->m_nNumClasses ? &Globals::Rules()->m_lstClasses[classId] : nullptr;
+    ASSERT_OR_THROW(pClass != nullptr);
+
+    if (auto* pCreature = creature(args))
```
Comment (Daztek) [inline Plugins/Creature/Creature.cpp]: You extract the arguments in the wrong order here. 

Should be
```
if (auto* pCreature = creature(args))
{ 
    const auto classId = Services::Events::ExtractArgument<int32_t>(args);
... 
```

File: Plugins/Creature/NWScript/nwnx_creature.nss [hunk 8]
```diff
@@ -1092,44 +1120,22 @@ void NWNX_Creature_SetAlignmentLawChaos(object creature, int value)
...
-
-    NWNX_PushArgumentInt(NWNX_Creature, sFunc, domain);
-    NWNX_PushArgumentInt(NWNX_Creature, sFunc, index);
-    NWNX_PushArgumentObject(NWNX_Creature, sFunc, creature);
-
-    NWNX_CallFunction(NWNX_Creature, sFunc);
+    return NWNX_Creature_SetDomain(creature, CLASS_TYPE_CLERIC, index, domain);
 }
 
 int NWNX_Creature_GetWizardSpecialization(object creature)
```
Comment (Daztek) [inline Plugins/Creature/NWScript/nwnx_creature.nss]: `WriteTimestampedLogEntry("WARNING: NWNX_Creature_GetWizardSpecialization() is deprecated, please use NWNX_Creature_GetSpecialization()");` also.

File: Plugins/Creature/NWScript/nwnx_creature.nss [hunk 9]
```diff
@@ -1092,44 +1120,22 @@ void NWNX_Creature_SetAlignmentLawChaos(object creature, int value)
 
 int NWNX_Creature_GetClericDomain(object creature, int index)
```
Comment (Daztek) [inline Plugins/Creature/NWScript/nwnx_creature.nss]: `WriteTimestampedLogEntry("WARNING: NWNX_Creature_GetClericDomain() is deprecated, please use NWNX_Creature_GetDomain()");` here too please!

**Comment by Daztek** [comment]
Also don't forget to update the changelog! :)

File: Plugins/Creature/Creature.cpp (hunk 7 above)
Comment (GoLoT) [reply Plugins/Creature/Creature.cpp]: Good catch, copying & pasting is always a bad idea.

**Comment by GoLoT** [comment]
Done. Added the extra deprecation warnings but kept the return statements for compatibility with existing code.

**Comment by GoLoT** [comment]
Added unit tests, all passed in my testing environment. Removing the untested tag.

**Pull request merged**